codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -341,4 +395,12 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
         super.handlerRemoved(ctx);
         cleanup();
     }
+
+    public final int blockSize() {
+        return blockSize;
+    }
+
+    public final int currentBlockLength() {
+        return currentBlockLength;
```

Comment:
OK, I've ditched `currentBlockLength` in favor of `buffer.readableBytes()`. This required a little clean up, which allowed me to make some of the code (especially the `while` loop in `encode()`) simpler and more understandable.

As we don't always write all the `msg`'s bytes that come into `encode()`, we only compress and write when the block size is reached (or, now on `flush()`). Thus, some data will remain buffered in the member field `buffer`. In `allocateBuffer()`, I now account for any of that buffered data when calculating the output buffer size. This was not a bug earlier because I did not set the max buffer size on the `allocateBuffer()`'s return value, and we are calling `out.ensureWritable()` later on (which would have incresed the buffer size).

Further, I've optimized `allocateBuffer()` to return an empty buffer if we know we do not have enough data between any previously not written data (still buffered in `buffer`) plus the incoming `msg`'s bytes. I would have preferred to return either `null` or `Unpooled.EMPY_BUFFER` but we call `release()` on the buffer after `encode()` is invoked from `MessageToByteEncoder#write()`.